enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,238 @@
+
+## Summary
+
+OpenShift should be capable of synchronizing user groups from a 3rd party
+identity provider upon a user logging in to their OpenShift cluster.
+
+## Motivation
+
+In the real world, OpenShift is not the owner of user identities but these
+are stored in a 3rd party identity provider system. When OpenShift retrieves
+such an identity, it usually only gathers information about the user's unique
+identification in that IdP and optionally a username that's preferred for the
+user if it's different from its identifier.
+
+It is a common practice to group users in logical groups on the identity provider
+side and sometimes it would make sense to bring the groups over to the OpenShift
+cluster as well. The identity providers oftentimes allow a protocol-specific path
+to retrieve group information, which is something OpenShift should leverage in those
+cases where it is desired.
+
+### Goals
+
+- allow retrieving user's groups from a 3rd party OIDC identity provider into
+  OpenShift when such a user logs in
+
+### Non-Goals
+
+- implement group synchronization for an identity provider type other than
+  OIDC
+- keep the groups retrieved by OpenShift from a 3rd party identity provider
+  synchronized between user logins
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+I am an administrator of an OpenShift cluster that uses identities from
+my company's identity provider. The users of the identity provider are
+organized in a complex and well-though-out group structure. I would like
+OpenShift to respect the group structure so that I can assign privileges
+to the users on OpenShift side according to their group membership in the
+identity provider.
+
+### Implementation Details
+
+#### OIDC Group Retrieval
+
+OIDC core specification does not mention a standard `id_token` claim that
+would serve for user group storage and so none of the standard scopes define
+accessing it. That means retrieving the groups will very likely always require
+the user to configure `extraScopes` in their identity provider configuration
+(`oauth.config.openshift.io/cluster`).
+
+A possible follow-up implementation might include being able to
+[request specific claims](https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter)
+from the UserInfo endpoint or from an `id_token`, which would allow for a more fine-grained
+configuration to reach the group claim.
+
+#### API
+
+The `OpenIDClaims` config API structure shall be expanded by a new `Groups`
+field:
+
+```golang
+// OpenIDClaims contains a list of OpenID claims to use when authenticating with an OpenID identity provider
+type OpenIDClaims struct {
+    ...
+
+    // groups is the list of claims whose values should be converted into OpenShift groups
+    // that will include the user
+    // +optional
+    Groups []string `json:"groups,omitempty"`
+}
+```
+
+The group object (`groups.user.openshift.io.com/v1`) should be annotated with
+`oauth.openshift.io/idp.<idpname>: "synced"`
```

Comment:
Perhaps we should also have an annotation with a timestamp, e.g. `oauth.openshift.io/idp.<idpname>.synctime: "<timestamp>"`? This could be useful when investigating synchronization issues between IdPs.

Reply:
Perhaps we could merge that into this annotation. Although I am a little hesitant because that might be too many updates to just the annotation itself and not really doing much to the object.

Reply:
The update may happen asynchronously in the background. I have no issues with that. 

The important question is do we want to update the timestamp on every ID Token rotation? In other words, what will happen when we exchange a Refresh Token for a new Access Token and get a new ID Token "by the way". This rotation will probably happen pretty often (Access Token lifetime set in Keycloak is 5 min by default).

Reply:
Clarified in other comments. We don't rotate those tokens.

Reply:
On a further thought, I would refrain from synctime in the annotations, purely to avoid updates that only update the annotation time.

Reply:
Could we create an event or an audit log for that instead? I believe this functionality might be crucial for sorting permission settings order problems in the future.